File: rusty_droid/app/src/main/java/de/knma/rustydroid/data/City.kt
package de.knma.rustydroid.data

data class City(
    val name: String,
    val population: Int,
    val area: Float,
    @get:JvmName("is_capital") val isCapital: Boolean,
    @get:JvmName("district_populations") val districtPopulations: IntArray,
    @get:JvmName("district_areas") val districtAreas: FloatArray
) {

    companion object {
        @JvmStatic
        @Suppress("FunctionName")
        external fun get_karlsruhe(): City
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as City

        if (population != other.population) return false
        if (area != other.area) return false
        if (isCapital != other.isCapital) return false
        if (name != other.name) return false
        if (!districtPopulations.contentEquals(other.districtPopulations)) return false
        if (!districtAreas.contentEquals(other.districtAreas)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = population
        result = 31 * result + area.hashCode()
        result = 31 * result + isCapital.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + districtPopulations.contentHashCode()
        result = 31 * result + districtAreas.contentHashCode()
        return result
    }
}

File: iron_bird_call/rust/src/api/city.rs
pub struct City {
    pub name: String,
    pub population: i32,
    pub area: f32,
    pub is_capital: bool,
    pub district_populations: Vec<i32>,
    pub district_areas: Vec<f32>,
}

pub fn get_karlsruhe() -> City {
    City {
        name: String::from("Karlsruhe"),
        population: 308436,
        area: 173.46,
        is_capital: false,
        district_populations: vec![21000, 18500, 15200, 32000, 28700, 19800, 24500, 17300],
        district_areas: vec![12.3, 8.7, 15.1, 22.4, 18.9, 11.2, 9.8, 14.6],
    }
}
